scripts/remove-unregistered-users.ts

```diff
@@ -0,0 +1,125 @@
+#!/usr/bin/env node -r esm -r ts-node/register
+
+import { resolve } from "path";
+
+import admin from "firebase-admin";
+
+import {
+  checkFileExists,
+  initFirebaseAdminApp,
+  makeScriptUsage,
+  parseCredentialFile,
+} from "./lib/helpers";
+
+const usage = makeScriptUsage({
+  description: `Bulk remove of user accounts that exist in the auth but are not in the firestore users collection`,
+  usageParams: `CREDENTIAL_PATH`,
+  exampleParams: `fooAccountKey.json`,
+});
+
+const [credentialPath] = process.argv.slice(2);
+if (!credentialPath) {
+  usage();
+}
+
+if (!checkFileExists(credentialPath)) {
+  console.error("Credential file path does not exists:", credentialPath);
+  process.exit(1);
+}
+
+const { project_id: projectId } = parseCredentialFile(credentialPath);
+
+if (!projectId) {
+  console.error("Credential file has no project_id:", credentialPath);
+  process.exit(1);
+}
+
+initFirebaseAdminApp(projectId, {
+  credentialPath: credentialPath
+    ? resolve(__dirname, credentialPath)
+    : undefined,
+});
+
+// 1000 is the maximum number of users firestore can fetch. check .listUsers()'s documentation for more information.
+// It is also 1000 by default but I decided to set it explicitly to avoid confusion.
+const MAX_USERS = 1000;
+
+const getAuthUsers = async (maxUsers: number, pageToken?: string) =>
+  await admin.auth().listUsers(maxUsers, pageToken);
+
+(async () => {
+  const totalAuthUsers: admin.auth.UserRecord[] = [];
+
+  console.log("Fetching firestore auth users...");
+  const { users: authUsers, pageToken } = await getAuthUsers(MAX_USERS);
+  totalAuthUsers.push(...authUsers);
+
+  // @debt This is going to make another request with offset if there are the maximum users returned by the previous request. This should be using recursion.
+  if (authUsers.length >= MAX_USERS) {
+    console.log(
+      `Auth users are more than ${MAX_USERS}. Fetching additional batch of firestore auth users...`
+    );
```

Comment:
Is this ever going to trigger? Presumably `authUsers.length` will be at most 1000, since that is the size that will be returned, and there will be some other method of checking if there are more users to be fetched?

Reply:
> Is this ever going to trigger?

Yes. I suppose I can leave it to `==` instead of `>=`. 

> and there will be some other method of checking if there are more users to be fetched?

There is no built-in method for it. Some people keep track of their users by incrementing a value in the database, which feels hacky. I suppose checking if the number of users fetched `==` the maximum that means there are more. 

Of course there is the edge case where the database can have exactly 1000 users and this logic will trigger again. I'm not sure if this will return an empty array or `undefined`.

Reply:
I think changing to `===` should be okay?

Reply:
Ah, yeah ok, makes sense. == is probably more obvious when reading the code, but maybe also a comment explaining this aspect of how we’re finding more + checking the docs to see what is returned when there are no more to get to ensure we handle that ‘final case’ correctly.

There’s other scripts that use this same SDK endpoint that seem to handle it in a loop, maybe check what they’re doing?